### ISSUE TITLE
test_warp_softmax_64bit_indexing_cuda_float16 takes ~147GB of CPU memory and is very slow

### ISSUE THREAD

**Issue by xwang233** [pytorch/pytorch#84944]
### 🐛 Describe the bug

test_warp_softmax_64bit_indexing_cuda_float16 takes ~147GB of CPU memory and is very slow

Reproduce:
```
/usr/bin/time -v python test/test_nn.py -v -k test_warp_softmax_64bit_indexing_cuda_float16
```

Output:
```
test_warp_softmax_64bit_indexing_cuda_float16 (__main__.TestNNDeviceTypeCUDA) ... ok

----------------------------------------------------------------------
Ran 1 test in 105.592s

OK
        Command being timed: "python test/test_nn.py -v -k test_warp_softmax_64bit_indexing_cuda_float16"
        User time (seconds): 776.69
        System time (seconds): 2031.00
        Percent of CPU this job got: 2576%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 1:48.96
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 147604400
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 184895836
        Voluntary context switches: 14118
        Involuntary context switches: 2191694
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

Considering the test was taking ~84 GB of ram before (https://github.com/pytorch/pytorch/pull/67922#issuecomment-963899593), but it takes 140 GB now, something could have gone wrong. Also, the test name was changed in https://github.com/pytorch/pytorch/pull/84182.

### Versions

torch_commit: https://github.com/pytorch/pytorch/commit/7a9ab5c232f54430704456d18a22f99838489817

V100 32 GB

cc @ptrblck @ngimel @malfet @eqy

**Comment by eqy** [comment]
I've bisected this to #67796 which changes the way `assertEqual` is implemented and could cause the additional memory usage (e.g., with an extra copy or something). I'll work on a workaround PR.

**Comment by ngimel** [comment]
Thanks for digging in!

**Comment by mruberry** [comment]
cc @pmeier -- maybe there's an opportunity to reduce memory usage?

**Comment by eqy** [comment]
Note that the softmax issue should be addressed in #85373, but not sure if we want to keep this open or use a new issue for investigating `assertEqual` memory usage.

**Comment by ppwwyyxx** [comment]
Ran into large memory footprint of `test_pdist_norm_large`. Part of the reason for the huge memory usage is due to the promotion to float64 at 
https://github.com/pytorch/pytorch/blob/a580a63448bf39c3a9783919c335990abe9e49be/torch/testing/_comparison.py#L1012-L1028

**Comment by ngimel** [comment]
cc @pmeier - why are we unconditionally promoting to fp64? This seems unnecessary.

**Comment by pmeier** [comment]
It was introduced in #60536 after we seen some problems with `torch.bool` inputs. TBH, I didn't pay attention to the memory footprint. IIRC, this cast needs to make sure that `is_close` works correctly:

https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/testing/_comparison.py#L991-L994

It seems that support for `torch.bool` was added to `isclose` in the mean time:

```pycon
>>> a = torch.tensor([0], dtype=torch.bool)
>>> b = torch.tensor([1], dtype=torch.bool)
>>> torch.isclose(a, a)
tensor([True])
>>> torch.isclose(a, b)
tensor([False])
>>> torch.isclose(a, b, atol=1)
tensor([True])
```

Since `isclose` also handles uint8 overflow

```
>>> a = a.byte()
>>> b = b.byte()
>>> abs(a - b)
tensor([255], dtype=torch.uint8)
>>> torch.isclose(a, b, atol=1)
tensor([True])
>>> torch.isclose(b, a, atol=1)
tensor([True])
```

I think we can safely remove the cast. However, we still need it for `torch.bool` and `torch.uint8` in case we need to provide diagnostics:

https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/testing/_comparison.py#L280-L284

I'll send a PR.

**Comment by ngimel** [comment]
Yeah leaving it only for bool and uint8 makes sense, and only for unhappy case of error makes sense.
